Sources/Datadog/DatadogInternal/Concurrency/Concurrency.swift

```diff
@@ -0,0 +1,56 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+
+/// A property wrapper using a fair, POSIX conforming reader-writer lock for atomic
+/// access to the value.
+///
+/// The wrapper is a class to prevent copying the lock, it creates and initilaizes a `pthread_rwlock_t`.
+/// An additional method `mutate` allow to safely mutate the value within a closure.
+@propertyWrapper
+public final class _pthread_rwlock<Value> {
+    /// The wrapped value.
+    private var value: Value
+
+    /// The lock object.
+    private var rwlock = pthread_rwlock_t()
```

Comment:
This read-write lock which allows many reader threads and a single writer thread at a time. Should be used when reading is done more often than writing. Seems like we are dealing with an opposite? 🤔

Reply:
Yes exactly, That's the case here: we will have few writes of tags and attributes, while we need available reads because we capture them on the caller thread when the user adds log entries.